channel.go

```diff
@@ -161,25 +166,45 @@ func (ch *Channel) send(msg message) (err error) {
 	return ch.sendOpen(msg)
 }
 
-func (ch *Channel) open() error {
-	return ch.call(&channelOpen{}, &channelOpenOk{})
+func (ch *Channel) open(ctx context.Context) error {
+	return ch.call(ctx, &channelOpen{}, &channelOpenOk{})
 }
 
 // Performs a request/response call for when the message is not NoWait and is
 // specified as Synchronous.
-func (ch *Channel) call(req message, res ...message) error {
+func (ch *Channel) call(ctx context.Context, req message, res ...message) error {
+	select {
+	case <-ctx.Done():
+		return ErrCanceled
+	default:
+	}
+
+	var status uint32
+
+	ch.responsesM.Lock()
+
 	if err := ch.send(req); err != nil {
+		ch.responsesM.Unlock()
 		return err
 	}
 
+	if req.wait() {
+		ch.responses = append(ch.responses, &status)
+	}
+
+	ch.responsesM.Unlock()
+
 	if req.wait() {
 		select {
+		case <-ctx.Done():
+			atomic.StoreUint32(&status, 1)
```

Comment:
There is a race conditions between receiving `ctx.Done()`, storing `status = 1` and loading `status` in `dispatch()`.

Instead, let's collect the in-flight contexts, so we will coordinate by communicating over `Context.Done()` rather than sharing the `status` state.

I think this would mean a responses slice of `[]atomic.Value` and storing and loading `context.Context` types.

Reply:
Alternatively, the responses slice could contain a buffered chan of messages and that would be used instead of `rpc`.